src/java.base/share/classes/java/lang/StringFormat.java

```diff
@@ -0,0 +1,350 @@
+                // no formatting to be done
+                return format;
+            }
+
+            int len = format.length();
+            if (off + 1 != len) {
+                int off1 = format.indexOf('%', off + 2);
+                String s = null;
+                if (args.length == 1) {
+                    if (off1 == -1) {
+                        s = format1(format, off, args[0]);
+                    }
+                } else if (args.length == 2) {
+                    if (off1 != -1 && off1 + 1 != len) {
+                        s = format2(format, off, off1, args[0], args[1]);
+                    }
+                }
+                if (s != null) {
+                    return s;
+                }
+            }
+        }
+
+        return new Formatter().format(format, args).toString();
+    }
+
+    private static String format1(String format, int off, Object arg) {
+        int len = format.length();
+        char conv = format.charAt(off + 1);
+        int width = 0;
+        if (conv >= '1' && conv <= '9') {
+            width = conv - '0';
+            if (off + 2 < len) {
+                conv = format.charAt(off + 2);
+            }
+        }
+
+        if (conv == STRING) {
+            if (isLong(arg)) {
+                conv = DECIMAL_INTEGER;
+            } else {
+                arg = String.valueOf(arg);
+            }
+        }
+
+        int size = stringSize(conv, arg);
+        if (size == -1) {
+            return null;
+        }
+        return format1(format, off, conv, arg, width, size);
+    }
+
+    private static String format2(String format, int off0, int off1, Object arg0, Object arg1) {
+        final int len = format.length();
+        char conv0 = format.charAt(off0 + 1);
+        int width0 = 0;
+        if (conv0 >= '1' && conv0 <= '9') {
+            width0 = conv0 - '0';
+            if (off0 + 2 < len) {
+                conv0 = format.charAt(off0 + 2);
+            }
+        }
+
+        char conv1 = format.charAt(off1 + 1);
+        int width1 = 0;
+        if (conv1 >= '1' && conv1 <= '9') {
+            width1 = conv1 - '0';
+            if (off1 + 2 < len) {
+                conv1 = format.charAt(off1 + 2);
+            }
+        }
+
+        if (off1 + 2 < len) {
+            if (format.indexOf('%', off1 + 2) != -1) {
+                return null;
+            }
+        }
+
+        String str;
+        byte coder = format.coder();
+        if (conv0 == STRING) {
+            if (isLong(arg0)) {
+                conv0 = DECIMAL_INTEGER;
+            } else {
+                str = String.valueOf(arg0);
+                coder |= str.coder();
+                arg0 = str;
+            }
+        }
+
+        if (conv1 == STRING) {
+            if (isLong(arg1)) {
+                conv1 = DECIMAL_INTEGER;
+            } else {
+                str = String.valueOf(arg1);
+                coder |= str.coder();
+                arg1 = String.valueOf(str);
```

Comment:
nit: `arg1 = String.valueOf(str);` can be avoided by just using `str`:

```suggestion
                arg1 = str = String.valueOf(arg1);
                coder |= str.coder();
```

Reply:
alternatively:
```suggestion
                str = String.valueOf(arg1);
                coder |= str.coder();
                arg1 = str;
```